src/lib/rnp.cpp

```diff
@@ -8174,29 +8174,32 @@ add_json_mpis(json_object *jso, pgp_key_t *key, bool secret = false)
 static rnp_result_t
 add_json_sig_mpis(json_object *jso, const pgp_signature_t *sig)
 {
-    pgp_signature_material_t material = {};
-    try {
-        if (!sig->parse_material(material)) {
-            return RNP_ERROR_BAD_PARAMETERS;
-        }
-    } catch (const std::exception &e) {
-        RNP_LOG("%s", e.what());
-        return RNP_ERROR_OUT_OF_MEMORY;
+    auto material = sig->parse_material();
+    if (!material) {
+        return RNP_ERROR_NOT_SUPPORTED;
     }
     switch (sig->palg) {
     case PGP_PKA_RSA:
     case PGP_PKA_RSA_ENCRYPT_ONLY:
-    case PGP_PKA_RSA_SIGN_ONLY:
-        return add_json_mpis(jso, "sig", &material.rsa.s, NULL);
+    case PGP_PKA_RSA_SIGN_ONLY: {
+        auto &rsa = dynamic_cast<const pgp::RSASigMaterial &>(*material);
+        return add_json_mpis(jso, "sig", &rsa.sig.s, NULL);
+    }
     case PGP_PKA_ELGAMAL:
-    case PGP_PKA_ELGAMAL_ENCRYPT_OR_SIGN:
-        return add_json_mpis(jso, "r", &material.eg.r, "s", &material.eg.s, NULL);
-    case PGP_PKA_DSA:
-        return add_json_mpis(jso, "r", &material.dsa.r, "s", &material.dsa.s, NULL);
+    case PGP_PKA_ELGAMAL_ENCRYPT_OR_SIGN: {
```

Comment:
IMHO this piece of code call for virtual function

something like 
```
virtual mpi* SigMaterial::get_mpi() = 0;
```
and appropriate redefinitions for ```EgSigMaterial``` et al

Reply:
I don't like the way it is currently done as well. The problem here is that different signature types may have different MPIs. In this case it is simple, one or two, but for EncMaterial and KeyMaterial it is more complicated, however I think we should use the same approach in both cases. Don't know a good solution yet.